src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,151 @@
+          'Access your data on %(count)s other sites',
+          count), { count: i18n.formatNumber(count) }
+        );
+      },
+    };
+    this.hostPermissions = [];
+  }
+
+  permissionStrings: Object;
+  hostPermissions: Array<React.Element<typeof Permission>>;
+
+  // Add individual Permission components for individual site/domain permissions.
+  formatItems(items: Array<string>, messageType: string) {
+    for (const item of items) {
+      this.hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.permissionStrings[messageType](item)}
+          key={item}
+        />
+      );
+    }
+  }
+
+  // Adds Permission components for a list of host permissions. If we have 4 or
+  // fewer, display them all, otherwise display the first 3 followed by an item
+  // that says "...plus N others".
+  addHostPermissions(
+    list: Array<string>,
+    singleMessageType:string,
+    moreMessageType:string
+  ) {
+    if (list.length < 5) {
+      this.formatItems(list, singleMessageType);
+    } else {
+      this.formatItems(list.slice(0, 3), singleMessageType);
+
+      const remaining = list.length - 3;
+      this.hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.permissionStrings[moreMessageType](remaining)}
+          key={moreMessageType}
+        />
+      );
+    }
+  }
+
+  render() {
+    const { permissions } = this.props;
+
+    // Group permissions into "site" and "domain" permissions. If any
+    // "all urls" permissions are found, break as we'll only need one
+    // host permissions message.
+    let allUrls = false;
+    const wildcards = [];
+    const sites = [];
+    for (const permission of permissions) {
+      if (permission === '<all_urls>') {
+        allUrls = true;
+        break;
+      }
+      if (permission.startsWith('moz-extension:')) {
+        continue;
+      }
+      const match = /^[a-z*]+:\/\/([^/]+)\//.exec(permission);
+      if (!match) {
+        log.debug(`Host permission string "${permission}" appears to be invalid.`);
+        continue;
+      }
+      if (match[1] === '*') {
+        allUrls = true;
+      } else if (match[1].startsWith('*.')) {
+        wildcards.push(match[1].slice(2));
+      } else {
+        sites.push(match[1]);
+      }
+    }
+
+    // Format the host permissions.  If we have a wildcard for all urls,
+    // a single string will suffice.  Otherwise, show domain wildcards
+    // first, then individual host permissions.
+    if (allUrls) {
+      this.hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.permissionStrings.allUrls}
+          key="allUrls"
+        />
+      );
+    } else {
+      this.addHostPermissions(wildcards, 'wildcard', 'tooManyWildcards');
+      this.addHostPermissions(sites, 'oneSite', 'tooManySites');
+    }
+    return (
+      <React.Fragment>
+        {this.hostPermissions}
```

Comment:
I think it would be better to have this be a variable created in the `render()` method that was modified in one function so it's easier to walk through. I think the side-effect style here/adding to the `hostPermissions` shared array is confusing.